strings/base_collections_base.h

```diff
@@ -189,64 +286,85 @@ WINRT_EXPORT namespace winrt
 
         void SetAt(uint32_t const index, T const& value)
         {
-            if (index >= static_cast<D const&>(*this).get_container().size())
+            static_cast<D&>(*this).perform_exclusive([&]
             {
-                throw hresult_out_of_bounds();
-            }
+                if (index >= static_cast<D const&>(*this).get_container().size())
+                {
+                    throw hresult_out_of_bounds();
+                }
 
-            this->increment_version();
-            static_cast<D&>(*this).get_container()[index] = static_cast<D const&>(*this).wrap_value(value);
+                this->increment_version();
+                static_cast<D&>(*this).get_container()[index] = static_cast<D const&>(*this).wrap_value(value);
```

Comment:
Good catch

Reply:
The latest iteration is a first stab at addressing this. I'm mostly looking for feedback/input on the approach, though it is fully functional. The most important thing is probably delaying destruction of the "wrapped" value vs. delaying destruction of the "unwrapped" value. Of course, delaying destruction of the "wrapped" value implies that the destruction of the "unwrapped"/"underlying" value is also delayed. The approach here is to delay destruction of the "wrapped" value. This seemed to be the safest approach, but is not without its pros/cons.

* This covers the scenario where the "wrapped" type's destructor does something potentially unsafe. I'm unsure how realistic this scenario is, however I figured it's best to be safe.
* We can make fewer assumptions about the "wrapped" type, at least compared to WinRT types. In particular how/if it's constructible. The approach here is to use `std::optional` so that we can delay (move) construct the value, and that is the only requirement on the type put in place. If the type is not move constructible, we fall back to running the destructor under the lock (a con, but unlikely to ever be a problem). There's also a potential issue if the moved-from value's destructor does something unsafe, however that's pretty much unavoidable no matter what we do.
* As-is, we assume that anything that's not trivially destructible needs to be destructed outside of the lock. This is a bit more restrictive than necessary. E.g. I would assume that `hstring` should be safe to destroy under the lock (not necessarily a con; doing less under the lock is generally good). Similarly, I would assume that structs with `IReference` members should be safe to destroy under the lock (but I'm not 100% sure, correct me if I'm wrong).
* This doesn't take into account the "threadedness" of the collection. I.e. the single threaded collections still try and order destruction. This is mostly noteworthy for operations like `Clear` and `ReplaceAll` where we may otherwise be able to re-use the buffer. This is mostly a simplicity decision.

Mostly just wanted to get this out to get thoughts/feedback.